### PR TITLE
Feature: Stackable NFTs Marketplace

### PR DESCRIPTION
**Pull request by chexware**

### CODE REVIEW

File: pallets/auction/src/lib.rs [hunk 1]
```diff
@@ -1080,6 +1102,69 @@ pub mod pallet {
 					<ItemsInAuction<T>>::insert(item_id, true);
 					Ok(auction_id)
 				}
+				ItemId::StackableNFT(class_id, token_id, amount) => {
+					// Check available balance
+					let available_balance = T::NFTHandler::get_stackable_nft_balance(&recipient, &(class_id, token_id));
+					let reserved_balance =
+						Self::reserved_stackable_nft_balances(recipient.clone(), (class_id, token_id));
+					ensure!(
+						amount + reserved_balance <= available_balance,
```
Comment (justinphamnz) [inline pallets/auction/src/lib.rs:1111]: This needs to be safe added to prevent overflow

File: pallets/nft/src/lib.rs [hunk 2]
```diff
@@ -1470,4 +1470,19 @@ impl<T: Config> NFTTrait<T::AccountId, BalanceOf<T>> for Pallet<T> {
 
 		Ok(minted_token_id)
 	}
+
+	fn get_stackable_nft_balance(who: &T::AccountId, asset_id: &(Self::ClassId, Self::TokenId)) -> BalanceOf<T> {
+		NftModule::<T>::get_stackable_collections_balances((asset_id.0, asset_id.1, who))
+	}
+
+	fn transfer_stackable_nft(
```
Comment (justinphamnz) [inline pallets/nft/src/lib.rs:1478]: As Reserved stackable nft has already implemented, we need to prevent it's being transferred as well.

**Review by justinphamnz — CHANGES_REQUESTED** [review]
Thanks @chexware. Please check the comment below

File: pallets/auction/src/tests.rs [hunk 3]
```diff
@@ -183,6 +207,43 @@ fn create_new_multicurrency_auction_bundle_work() {
...
+				start: 1,
+				end: Some(101),
+			})
+		);
+		assert_eq!(
+			AuctionModule::items_in_auction(ItemId::StackableNFT(0, 0, 50u128)),
+			Some(true)
+		);
+		assert_eq!(Balances::free_balance(ALICE), 99996);
+		assert_eq!(AuctionModule::reserved_stackable_nft_balances(ALICE, (0, 0)), 50u128);
```
Comment (justinphamnz) [inline pallets/auction/src/tests.rs:243]: Also need to test NFT is not able to be transferred.

File: pallets/auction/src/lib.rs (hunk 1 above)
Comment (chexware) [reply pallets/auction/src/lib.rs]: should be resolved in the  following PR

File: pallets/nft/src/lib.rs (hunk 2 above)
Comment (chexware) [reply pallets/nft/src/lib.rs:1478]: should be resolved in the following commit

Comment (chexware) [reply pallets/nft/src/lib.rs:1518]: fixed in the latest commit

File: pallets/auction/src/lib.rs [hunk 4]
```diff
@@ -1080,6 +1096,72 @@ pub mod pallet {
 					<ItemsInAuction<T>>::insert(item_id, true);
 					Ok(auction_id)
 				}
+				ItemId::StackableNFT(class_id, token_id, amount) => {
+					// Check available balance
+					//let available_balance = T::NFTHandler::get_stackable_nft_balance(&recipient, &(class_id,
```
Comment (justinphamnz) [inline pallets/auction/src/lib.rs:1101]: If it's not usable, we should remove the commented out code

**Review by justinphamnz — CHANGES_REQUESTED** [review]
Looks great. Thanks @chexware. 

We need to do some storage migration as ItemId changes that change the storage structure of item index

File: pallets/auction/src/lib.rs (hunk 4 above)
Comment (chexware) [reply pallets/auction/src/lib.rs]: just removed it. It was left from the old implementation

**Review by justinphamnz — APPROVED** [review]

**Pull request merged**